### ISSUE TITLE
Re-estimating read at phylum rank comes out very strange

### ISSUE THREAD

**Issue by larssnip** [jenniferlu717/Bracken#103]
I have a kraken2 database with the standard libraries for human, bacteria, archaea, viruses, fungi and plants, i.e. the four domains (superkingdoms) Eukaryota, Bacteria, Archaea and Viruses are represented. I have a kraken2 report from a sample. If I use bracken to re-estimate at the species rank, and then sum reads allocated to the four domains listed above, the results seems sensible, and comparable to the raw kraken2 results. The same happens if I use bracken to re-estimate at genus, family or order ranks. The results summed at the domain rank show more or less the same for all four domains. But, using bracken to re-estimate at class or phylum rank, the archaea and virus results are more or less swapped! This sample has around 3.7 million reads from viruses, and only a few hundred from archaea if we believe the kraken2 results, as well as the first bracken results (lower rank re-assignments). But, when bracken re-estimates at the high ranks, the archaea come out with 3.7 million reads instead! How is this at all possible? The input is the same, the databases are the same, the only difference is the rank at which I ask bracken to re-estimate. Anyone seen this before?

**Comment by jenniferlu717** [comment]
I think it has to do with a recent change I made that pushes all reads down to the genome level and then reestimates back up. Can you send me your Kraken report,Bracken results and the kmer distribution file (jennifer.lu717@gmail.com)

**Comment by larssnip** [comment]
OK, here they are...
[bl3_class.txt](https://github.com/jenniferlu717/Bracken/files/4753582/bl3_class.txt)
[bl3_family.txt](https://github.com/jenniferlu717/Bracken/files/4753583/bl3_family.txt)
[bl3_genus.txt](https://github.com/jenniferlu717/Bracken/files/4753584/bl3_genus.txt)
[bl3_kraken.txt](https://github.com/jenniferlu717/Bracken/files/4753585/bl3_kraken.txt)
[bl3_order.txt](https://github.com/jenniferlu717/Bracken/files/4753586/bl3_order.txt)
[bl3_phylum.txt](https://github.com/jenniferlu717/Bracken/files/4753587/bl3_phylum.txt)
[bl3_species.txt](https://github.com/jenniferlu717/Bracken/files/4753588/bl3_species.txt)

[bracken.zip](https://github.com/jenniferlu717/Bracken/files/4753597/bracken.zip)

**Comment by larssnip** [comment]
...the zip archive contains also the kmer distribution file...

**Comment by jenniferlu717** [comment]
I figured it out. Most of the Viral reads are in this subtree:

`3.78  2270015 3   D   10239     Viruses
3.78  2265073 0   O   28883       Caudovirales
3.78  2265059 4   F   10699         Siphoviridae
3.77  2264800 0   G   1921122         Nonanavirus
3.77  2264800 0   G1  2315205           unclassified Nonanavirus
3.77  2264800 2264800 S   1969841             Proteus phage VB_PmiS-Isfahan
`

But if you notice, there is no phylum. What happens is that Bracken assumes that the phylum printed before the species is the "parent" phylum for that species. Before the Viruses section in the report is the Archaea section. I've never seen this before though, where it goes straight from the domain level to order. 

Is there no phylum or class for that subtree? I can try to make a change to have it ignore those reads, but I also find it suspicious that there isnt even a phylum...

**Comment by Wrzlprmft** [comment]
@jenniferlu717: This sounds similar to the issue I reported [here](https://github.com/jenniferlu717/Bracken/issues/102#issuecomment-640784243).

My naïve approach would be to assign reads to the next higher level if the target level is missing. For example, if there is no phylum, report the respective reads at the kingdom level (or since that is missing here as well, to domain). As a user, I definitely prefer a correct result over one with strictly homogeneous levels.

**Comment by larssnip** [comment]
What is a "correct" result in this case? All reads being counted?

I would be confused if I asked bracken to estimate read counts at the phylum rank, and got the answer to some other rank instead. Then, at least, the output should indicate this, that the requested rank was absent from the taxonomy. The latter is perhaps always a good idea, to issue some warning if the requested rank is absent from some clade?

**Comment by Wrzlprmft** [comment]
> What is a "correct" result in this case? All reads being counted?

Yes, or more precisely no identifiable reads being discarded. At least I usually want to primarily know the distribution of the reads in the sample. If this is not correct, results can be easily misleading. Therefore I think this should have priority over the desired level.

For an extreme example, if I have data from an almost exclusively bacterial sample, estimating the distribution at the kingdom level could tell me that it’s 100% plants because there are no kingdoms of bacteria and there is one plant cell that somehow made it into my sample.

A more realistic example would be the `S1` level (subspecies). If I use this, I simply want Bracken to go as deep as possible, which is relevant if I have extremely diverse species like *E. coli* or *P. aeruginosa* in my sample. However, right now, I cannot use the results directly because the database often stops at the species level.

**Comment by larssnip** [comment]
Well, I think there will be differing opinions on this. If I have a million reads, and bracken tells me only 100 of them matches at the (plant) kingdom, then I think I would have a look at the kraken2 results they are based on, and quickly find there are plenty of bacteria in there. Always inspect the kraken results as well...

Identifiable reads are also highly dependent on the kraken database itself. We sequenced some gut microbiome from deer, and found they contained a lot of "human" reads, i.e. beware of carnivore bambi! It was of course deer-DNA that was assigned to human, since the latter was in the database but deer was not.

**Comment by Wrzlprmft** [comment]
Sure, I wouldn’t blindly accept a result like that. However, this may cause unnecessary effort. Here, I would have look at the domain level to get a broad overview of my sample.

Again, for a more practical example, consider the `S1` example I gave. To get full results at the highest depth, I would have to manually merge the results for `S` and `S1`.

**Comment by larssnip** [comment]
Yes, but this is not really a bracken problem, is it? The problem is that NCBI do not assign tax_id's below the species (S) anymore? Some genomes have an S1 tax_id, but many don't. Going below species rank is clearly a challenge, but new ideas pop up from time to time...

How would you merge the results for S and S1, if it is the S1 results you are interested in? Let us say species S has 10 genomes in the database. Three of them have S1 tax_id's, while seven have only the S tax_id. Kraken readcounts for the three S1's we call x1, x2 and x3, and the aggregated readcount for S is y.

**Comment by Wrzlprmft** [comment]
> The problem is that NCBI do not assign tax_id's below the species (S) anymore?

Don’t they? Either way, given that the species level is somewhat arbitrary and dominated by historic momentum for non-sexually reproducing organisms, it simply does not make sense to stop at the same depth everywhere.

> How would you merge the results for S and S1, if it is the S1 results you are interested in? Let us say species S has 10 genomes in the database. Three of them have S1 tax_id's, while seven have only the S tax_id. Kraken readcounts for the three S1's we call x1, x2 and x3, and the aggregated readcount for S is y.

This case should be tricky for Bracken anyway and I admit that I don’t know how it is handled right now. Anyway, my issue is rather with species that have substructure and those who don’t. Right now, I would use the Bracken result for the `S` level and then replace the entries for species with a substructure by the respective `S1` results.

**Comment by Wrzlprmft** [comment]
To give another specific example, here is a specific example where things go extremely wrong when applying Bracken at the `S1` level. This is an excerpt from a Kraken report (everything on the *P. aeruginosa* subgenus):

```
1006790	484	G1	136841	                Pseudomonas aeruginosa group
1003073	275991	S	287	                  Pseudomonas aeruginosa
630184	630184	S1	1407649	                    Pseudomonas aeruginosa RB
37312	37312	S1	1457194	                    Pseudomonas aeruginosa CF_PA39
18756	18756	S1	1009714	                    Pseudomonas aeruginosa PAK
…	…	S1	…	                    [twenty other P. aeruginosa subspecies]
1216	1216	S	198616	                  Pseudomonas jinjuensis
611	611	S	53408	                  Pseudomonas citronellolis
449	4	G2	627141	                  Pseudomonas nitroreducens/multiresinivorans group
405	405	S	95301	                    Pseudomonas multiresinivorans
40	0	S	46680	                    Pseudomonas nitroreducens
40	40	S1	1215112	                      Pseudomonas nitroreducens NBRC 12694
349	0	S	53412	                  Pseudomonas resinovorans
349	349	S1	1245471	                    Pseudomonas resinovorans NBRC 106553
191	191	S	29435	                  Pseudomonas flavescens
132	0	G2	1232139	                  Pseudomonas oleovorans/pseudoalcaligenes group
78	78	S	1149133	                    Pseudomonas furukawaii
54	28	S	301	                    Pseudomonas oleovorans
17	17	S1	1215116	                      Pseudomonas pseudoalcaligenes NBRC 14167
9	9	S1	1182590	                      Pseudomonas pseudoalcaligenes CECT 5344
110	110	S	43263	                  Pseudomonas alcaligenes
70	16	S	300	                  Pseudomonas mendocina
34	34	S1	399739	                    Pseudomonas mendocina ymp
20	20	S1	1197721	                    Pseudomonas mendocina DLHK
60	60	S	53406	                  Pseudomonas anguilliseptica
35	35	S	47882	                  Pseudomonas straminea
10	0	S	521720	                  Pseudomonas caeni
10	10	S1	521719	                    Pseudomonas caeni DSM 24390
```

And here is the corresponding Bracken report:

```
13965871	0	G1	136841	                Pseudomonas aeruginosa group
11658840	0	S	287	                  Pseudomonas aeruginosa
10849853	10849853	S1	1407649	                    Pseudomonas aeruginosa RB
48779	48779	S1	1457194	                    Pseudomonas aeruginosa CF_PA39
194403	194403	S1	1009714	                    Pseudomonas aeruginosa PAK
…	…	S1	…	                    [twenty other P. aeruginosa subspecies]
2305946	0	S	521720	                  Pseudomonas caeni
2305946	2305946	S1	521719	                    Pseudomonas caeni DSM 24390
712	0	S	53412	                  Pseudomonas resinovorans
712	712	S1	1245471	                    Pseudomonas resinovorans NBRC 106553
216	0	S	300	                  Pseudomonas mendocina
60	60	S1	399739	                    Pseudomonas mendocina ymp
156	156	S1	1197721	                    Pseudomonas mendocina DLHK
103	0	G2	627141	                  Pseudomonas nitroreducens/multiresinivorans group
103	0	S	46680	                    Pseudomonas nitroreducens
103	103	S1	1215112	                      Pseudomonas nitroreducens NBRC 12694
51	0	G2	1232139	                  Pseudomonas oleovorans/pseudoalcaligenes group
51	0	S	301	                    Pseudomonas oleovorans
51	51	S1	1215116	                      Pseudomonas pseudoalcaligenes NBRC 14167
```

Note how the counts for *P. caeni* were inflated from 10 to 2305946. I don’t think this is a desirable result under any circumstance. (I can e-mail the full files if desired.)

**Comment by larssnip** [comment]
Simple question: What was the target rank specified to bracken here? Was it S1?

**Comment by Wrzlprmft** [comment]
Yes.

**Comment by Wrzlprmft** [comment]
To suggest a solution that is conceptually simple and easy to communicate, but probably somewhat nasty to implement: Insert monotypic dummy taxa to inter- or extrapolate all branches of the phylogenetic tree to the target level and treat them like their parent taxon when reporting.
